src/generator/codegen.jl

```diff
@@ -482,11 +482,44 @@ function emit!(dag::ExprDAG, node::ExprNode{<:AbstractStructNodeType}, options::
     expr = Expr(:struct, false, struct_sym, block)
     field_cursors = fields(getCursorType(node.cursor))
     field_cursors = isempty(field_cursors) ? children(node.cursor) : field_cursors
-    for field_cursor in field_cursors
+
+    current_offset = 0
+    pads = []
+    for (i, field_cursor) in enumerate(field_cursors)
         field_sym = make_symbol_safe(name(field_cursor))
         field_ty = getCursorType(field_cursor)
+        field_offset = getOffsetOfField(field_cursor) ÷ 8
+        field_size = getSizeOf(field_ty)
+        @assert field_offset >= current_offset
+
+        if field_offset != current_offset
+            pad_size = field_offset - current_offset
+            pad_name = Symbol("pad", length(pads)) |> gensym
+            push!(block.args, Expr(:(::), pad_name, :(NTuple{$pad_size, UInt8})))
+            push!(pads, (;narg = i + length(pads), size = pad_size))
+        end
+
+        current_offset += field_size
```

Comment:
This fix only handles generic cases, the same fix should also apply for `StructMutualRef`.